userland/examples/ble_passive_scanning/main.cpp

```diff
@@ -0,0 +1,54 @@
+#include <cstring>
+#include <forward_list>
+#include <simple_ble.h>
+#include <stdio.h>
+#include <string.h>
+
+#include "advertisement.h"
+#include "advertisement_list.h"
+
+/*
+ * BLE Demo Application
+ * Passive scanner for Bluetooth Low Energy advertisements
+ */
+
+
+const int BUF_SIZE = 39;
+static unsigned char scan[BUF_SIZE];
+AdvertisementList list;
+
+static void callback(int result, int len, __attribute__((unused)) int unused2,
+                     __attribute__((unused)) void* ud)
+{
+  if (result == TOCK_SUCCESS) {
+    if (Advertisement::validAdvertisement(scan, len)) {
+      Advertisement adv(scan, len);
+
+      // new device detected
+      if (!list.containsDevice(adv)) {
+        if (list.add(adv)) {
+          list.printList();
+        }
+      }
+
+      // FIXME: add this to get dynamic behavior i.e, update every time new advertisement is detected
+      // but might it fill the print buffer, use at your own risk
+      // else {
+      //   list.newData(adv);
```

Comment:
Does std have a set datastructure? Could use that instead, keyed by the advertiser's address, so that in realistic situations, the set wouldn't grow arbitrarily large: new advertisements would always replace old ones from the same advertiser

Reply:
Yes, cpp has both set and hashmap but these need more memory (I tested with std::vector way way back).

I think a a hashmap is probably even better for this use case. Regardless I used ```forward_list``` because it should require as much memory as its C counter-part.